### ISSUE TITLE
Training details

### ISSUE THREAD

**Issue by zichunxx** [gemcollector/maniwhere#10]
Hi, zhecheng!

Recently, I tried to apply the contrastive loss in your paper to my training tasks, but I found some problems. Hope to discuss with you.

1. When the calculation of the contrastive loss is initiated after a latency, the performance curves (reward or success rate) degrade drastically. Sometimes the curve can rise back through subsequent training, sometimes it can't. Is this situation the same for your training processes? If so, are increasing the latency frames and decreasing the contrastive loss weights optional methods?

2. I think camera position has a potential but important impact on training an ideal visual encoder. For some small objects, the camera position needs to be adjusted to see more details of the object, but the ability to perceive the global environment is also lost, which is limited by the image size due to the huge training cost. Have you found a similar phenomenon, i.e., the effect of the main camera position on the training process?

Looking forward to your reply! Thanks!

**Comment by gemcollector** [comment]
Hi zichun! Thanks for your interest! I apologize for the delayed response.  
1. yea, actually, we tune the hyper-parameters of contrastive terms carefully in case the whole training breaks down.  In this version of maniwhere, it is more stable. You can try to decrease the contrastive loss weights to alleviate the problem.
2. In our setting, I think our two global cameras are enough to solve these tasks. The camera viewpoints were sampled randomly; we didn’t deliberately choose specific angles. Our only constraint was that both the robot and the object had to be visible in the frame. Which “small objects” setting are you referring to?

**Comment by zichunxx** [comment]
Thanks for your kind reply, zhecheng!

1. I'm still confused about when to start calculating the contrastive loss. If launched with a short latency, the model seems not to adapt to the current visual distribution. Additional loss is more likely to induce oscillations. Conversely, if a longer delay is set, the contrastive loss will be large when initiated, and the model will also be prone to collapse. Which way is better?

2. Taking `UR5_lift` as an example, I want to observe more details of the box to facilitate the convergence and move the camera closer to the table, but this comes at the cost of missing a full view of the manipulator. I found this will introduce more uncertainty about the subsequent training of the contrastive loss. The encoder is more likely to be attracted by the surrounding environment.

**Comment by gemcollector** [comment]
Thanks for your question!
1. I recommend delaying the contrastive loss. My principle to set the latency is to let the policy almost succeed in finishing the task, so that the contrastive loss will not be prone to collapse. 
2. Sounds interesting. So your two views are: one local view and one global view? And you find that the local view will make the training unstable?

**Comment by zichunxx** [comment]
Yes, The local view does lead to more fluctuations in training. 

One last question hope you can answer. I tried to reproduce the results of `ur5_lift`. Here is the feature map of the last encoder layer after 300K training frames:

![Image](https://github.com/user-attachments/assets/bf6b96f8-5bd7-4f70-839e-6ef07b9a0961)

The contrastive loss makes the encoder concentrate more on image boundaries. Have you seen this before? Thanks!

**Comment by gemcollector** [comment]
I think we should first clarify one point: what do you mean "feature map" here? Did  you directly visualize the "feature map"  (B, C, H, W)? Could we refer the "feature map" as the encoder concentration?

**Comment by zichunxx** [comment]
I visualized the feature maps (B, C, H, W) directly. I gave an inappropriate description above. But the feature map I got above doesn't present the contours of the arm and the rest of the scene as in your paper. So, I interrupted this training process to check for potential errors. I'm not sure if the feature map is normal when the contrastive loss is considered. Before that, the feature map was displayed as expected.

**Comment by gemcollector** [comment]
I see, there might be two reasons I think:
1. I visualize the attention when the policy is finished training. So you just visualize the feature map during training?
2. If the model can get a high reward, but the feature maps are visualized as above, I recommend you to use Grad CAM to show the attention. I think feature maps cannot fully represent the attention.

**Comment by zichunxx** [comment]
Thanks for your patience. 

1. The attention you mentioned means the feature map? I did visualize the feature map to supervise the training process.
2. Okay. I will try Grad CAM to show the encoder concentration.

**Comment by gemcollector** [comment]
So you add another "attention" loss to the training pipeline?

**Comment by zichunxx** [comment]
I just visualize feature maps by periodically saved snapshots. No additional loss is added.

**Comment by gemcollector** [comment]
I see, you can compare the feature map between the final policy with the intermediate policy. I am looking forward to seeing sth interesting

**Comment by zichunxx** [comment]
Thanks, zhecheng. I will try to finish one whole training process first.

**Comment by zichunxx** [comment]
Hi! I completed the training process with 700K frames and discovered that there were still some issues with the visualization of the Grad-CAM:

<img width="417" alt="Image" src="https://github.com/user-attachments/assets/c7d8a516-6fad-4b9e-9bbf-cca731637b7c" />

However, the episode reward has almost converged to a stable level:

![Image](https://github.com/user-attachments/assets/361a9827-0327-42cc-be7f-a8853d8888e8)


Here is the relevant code that requires the Grad-CAM with `pytorch-grad-cam`:

```
  original_image = time_step.observation[self._obs_channel:][:3,:,:].transpose(1, 2, 0) / 255
  input_tensor = torch.from_numpy(time_step.observation[self._obs_channel:]).unsqueeze(0).to(self.device)

  target_layers = [self.agent.encoder.model.layer2]

  cam = GradCAM(model=self.agent.encoder, target_layers=target_layers)

  grayscale_cam = cam(input_tensor=input_tensor, targets=None, aug_smooth=False, eigen_smooth=True)
  grayscale_cam = grayscale_cam[0, :]
  cam_image = show_cam_on_image(original_image, grayscale_cam, use_rgb=True)
```

Could you please give me some advice on potential issues? Thanks!
